Grouped svg mutations

hyanwong opened the pull request:
This follows on from #570 and picks up the idea of creating an SVG group for each mutation too (see https://github.com/tskit-dev/tskit/pull/570#issuecomment-624011266 for justification).

It seems like this is a good idea to me. The main thing to sort out is how we assign classes to the groups. At the moment I assign the parent node id (or "root") plus the child node id to all groups, including mutation groups. Mutation groups then have an additional 'mut" class and the mutation number and site number, whereas node groups have a "node" class. An example would be something like this (assuming an example with mutation 0 over a root node id 8 with 2 daughters, id 4 and 5)

```xml
<g class="tree t0">
  <g class="root n8 mut m0 s0">  <!-- a mutation group identifying mutation 0 over node 8 -->  
    <g class="root n8 node">   <!-- a node group identifying node 8-->
      <g class="p8 n4 mut m1 s1">  <!-- a mutation group identifying mutation 1 over child node 4 -->
      ... further mutation or node groups, including the node group for node 4 and descendants
      </g>
      <g class="p8 n5 mut m2 s2">  <!-- a mutation group identifying mutation 2 over child node 5 -->
      ... further mutation or node groups, including the node group for node 5 and descendants
      </g>
      <path class="edge" ...> <!-- the line joining the root node to mutation 0 -->
      <circle ...> <!-- the root node symbol, doesn't strictly need a class, as `edge + *` will do -->
      <text ...> <!-- the root node label, always a text node -->
    </g>
    <path class="edge" ...> <!-- the line above mutation 0 -->
    <rect ...> <!-- the mutation symbol for mutation 0 -->
    <text ...> <!-- the label for mutation 0, always a text node -->
  </g> <!-- close mutation group -->
</g> <!-- close tree group -->
```

This is all quite complicated to describe, but logical. I think it allows us to target anything we might want, but it could be easier to explain if, for instance, the circles and rects are given a class "symbol", rather than targetted because they always follow an element of class "edge".

Probably easiest if I discuss this with someone (e.g. @benjeffery ) on a Jitsi call?

benjeffery commented:
We just had a call where we thought the symbols and labels should have a class as the HTML entity may change and shouldn't be used as a target.

hyanwong commented:
I wonder if we should also add the population id and individual id to the node group (perhaps only if these are not `tskit.NULL`)? The obvious notation would be

```
<g class="pPOPNUM iIND_NUM ... p8 n4 mut m1 s1">
```

but `pID` for population would conflict with `pID` used to indicate parent ID. My inclination is to change the notation for parent nodes. Perhaps `aID` for "ancestorID" or `mID` for "motherID" (and `m` is the letter before `n`, so we would have `mA nB` for parent A + daughter B.

petrelharp commented:
> I wonder if we should also add the population id and individual id to the node group (perhaps only if these are not tskit.NULL)?

That'd make implementing the `.draw()` method even easier! Maybe `pop<POP ID>` and `ind<IND ID>`?

jeromekelleher commented:
The more semantic info we're including the better, I guess. Seems like there's no harm in it? We can always provide methods to trim out the info later if it turns out to add too much weight.

hyanwong commented:
I've rebased, implemented populations and individuals for nodes ("p<POP_ID>" and "i<IND_ID>", with parent nodes being "a<PARENT_NODE_ID" - other alternative letters welcome, although given the number of times these are repeated, I've stuck to single letters). So this is also ready for review now - the final piece of my SVG jigsaw! 

For discussion is Peter's suggestion about whether to by default create a UID for the SVG, to stop user-generated styles spilling into different SVGs in the same document (which I agree is very unexpected, and can play havoc with custom SVGs in Jupyter notebooks). However, I can't think of an elegant way to do this automatically: the easiest way is to require the user to specify an ID in the style specification, which would mean they would have to extract any autogenerated ID anyway, unless we defined some sort of placeholder. Alternatively, we could require the style specifications to be an array of strings, and we automatically prefix each string with "#UID " when writing to the SVG file, but messing with the user-supplied styles like this seems a little wrong.

Finally, I was also thinking that it would be useful to embed the number of mutations, sites, nodes, and edges per tree as metadata into the SVG somehow, as these are useful for animation. However, the most obvious way to add them as [`data-*` attributes](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/data-*) is SVG2 only, and not supported by svgwrite. We could possible define our own namespace for this (see the top answer at https://stackoverflow.com/questions/15532371/do-svg-docs-support-custom-data-attributes) but that seems quite heavyweight. Perhaps @benjeffery has a better solution?

When all seems OK, I'll write some docs and amend the commit with those.

benjeffery commented:
For the UID issue - I don't think we want to be too clever here with re-writing styles etc. I think a note in the docs to add a UID to both your styles and top-level attributes is fine. 

All browsers support data attributes: https://caniuse.com/#feat=mdn-svg_attributes_data and `svgwrite` can be persuaded to support any attribute you like:

```python
import svgwrite
svgwrite.data.full11.elements['g'].valid_attributes = frozenset(
    list(svgwrite.data.full11.elements['g'].valid_attributes) + 
    ['data-whatever']
)
svgwrite.data.full11.attributes['data-whatever'] = svgwrite.data.types.SVGAttribute(
    'data-whatever', anim=False, types=frozenset(['number']), const=[]
)
d = svgwrite.Drawing()
d.add(d.g(data_whatever=5))

You need the first two lines just after the import.

jeromekelleher commented on python/tskit/drawing.py:
```diff
@@ -253,10 +254,15 @@ def __init__(
         dwg = svgwrite.Drawing(size=self.image_size, debug=True, **root_svg_attributes)
         self.drawing = dwg
         dwg.defs.add(dwg.style(SvgTree.standard_style))
+        self.node_labels = {u: str(u) for u in range(ts.num_nodes)}
```
self.node_labels is being set twice, I think.

jeromekelleher commented on python/tskit/drawing.py:
```diff
@@ -577,38 +606,89 @@ def assign_x_coordinates(self, tree, x_start, width):
         if v == NULL:
-            classes.add(f"root")
+            classes.append(f"root")
+            if self.root_branch:
```
I don't think this if statement does anything - you've computed the ``self.root_branch_length`` above so there's no need for the local variable.

jeromekelleher commented on python/tskit/drawing.py:
```diff
@@ -577,38 +606,89 @@ def assign_x_coordinates(self, tree, x_start, width):
+                )
+            )
+            # after the first sideways line of an edge all further movements go downwards
+            offset_x = edge_x = 0
```
Avoid these multiple assignments.

jeromekelleher commented on python/tskit/drawing.py:
```diff
@@ -577,38 +606,89 @@ def assign_x_coordinates(self, tree, x_start, width):
+        # "sample":       a class present if the focal node is a sample
+        # "leaf":         a class present if the focal node is a leaf
+        classes.append("node")
+        ind = self.tree.tree_sequence.node(focal).individual
```
Not much point in looking up the node twice 
```
focal_node = self.tree.tree_sequence.node(focal)
if focal_node.individual != NULL:
   # etc
```

jeromekelleher commented on python/tskit/drawing.py:
```diff
@@ -617,73 +697,59 @@ def draw(self):
-                dx = pv[0] - pu[0]
-                dy = pv[1] - pu[1]
+            dx, dy = curr.edge_dxy
+            if dx == dy == 0:
```
Ouch, I didn't even realise that python let you do this. ``if dx == 0 and dy == 0``

jeromekelleher replied:
I doubt this does what you expect, either, so we're probably missing some test coverage. This should mean (dx == dy) == 0 -> (True | False) == 0 .

hyanwong replied:
I'm pretty sure a==b==c does the expected thing due to [comparison operator chaining](https://docs.python.org/3/reference/expressions.html#comparisons); "unlike C, expressions like a < b < c have the interpretation that is conventional in mathematics"

So 0==0==1 is False, and all works as expected. Personally I find this syntax clearer, but then my head is rarely in C mode.

benjeffery replied:
I prefer `if dx == 0 and dy == 0` as it is unambiguous for both "C thinkers" and "mathematics thinkers".

hyanwong replied:
Sure, happy to switch: just pointing out that it's not unreasonable.

hyanwong replied:
(by the way, does that mean you also dislike `a<b<c`, which I fund much clearer than breaking up the test into 2 halves, like you have to do in C)?

jeromekelleher replied:
You're right, looks like ``0 == 0 == 0`` is True rather than False. But, it's not obvious, so I would be against using it. Stuff like ``a < b < c`` is great, on the other hand, as the meaning is unambiguous (to my eyes anyway).

hyanwong commented:
I'm not paying much attention to this for the next week or two, but just to note that the main point of this PR, which assigns part of the line that represents an edge to a `mut` SVG group, means that when you try to animate an SPR between two trees, it's hard to find corresponding edges to transform, because the single line representing an edge may be broken down into pieces.

However, I *think* it might be possible to have the best of both worlds here, by drawing the line from a true node (not a mutation) upwards through the mutations and back to the previous node. I've coded this up in f365d51 but it will need a little thought/discussion, hopefully in a few weeks time.

jeromekelleher commented:
Hi @hyanwong, we're hoping to get 0.3.0 out in a couple of weeks and we should get this finished first, I guess? How much is left to do here?

hyanwong commented:
So I stopped developing this @jeromekelleher because I realised that it might make our lives more difficult if we want to map one tree to an adjacent tree in the tree sequence. With the "mutation makes an SVG group" idea, we no longer maintain some identical groups when moving to a new tree, so e.g. animating SPRs becomes more problematic.

I think we need some discussion on the right approach here. @benjeffery normally has useful things to say.

jeromekelleher commented:
I think it's probably OK to leave this go until the next release, unless you think there are backwards incompatible changes that will need to be made later to accommodate mutations? Do you foresee the structure of the SVG changing?

hyanwong commented:
Well, *if* we want to create a new subtree group under each mutation (i.e. as implemented in this PR) then yes, the SVG structure will be quite different. Sorry

jeromekelleher commented:
> Well, if we want to create a new subtree group under each mutation (i.e. as implemented in this PR) then yes, the SVG structure will be quite different. Sorry.

Hmmm. Well, I'm going to make this your call then. We do want to ship in the next few weeks, so it would be good to have a think about it. 

We could version the SVG output, so at least we could tell users about this, and this would let us make changes without having to worry *too* much about compatability?

jeromekelleher commented:
I've tagged this for 0.3.1, as it's pretty clear we're not going to get it sorted for 0.3.0. I think we'll just have to mark this SVG structure as "in development" and warn people to expect breakage if they depend on the structure.

hyanwong commented:
Yes - I think we might want not to implement SVG groups for mutations anyway. I'm debating whether to close this issue and try a new approach. We should probably have a more general discussion some time.

hyanwong commented:
Closing for the time being, as we are unsure if grouped mutations will be quite as flexible as we had hoped
